src/element/element_stream_writer.rs

```diff
@@ -0,0 +1,697 @@
+        let annotations = self.take_current_annotations();
+        let field_name = self.field_name.take();
+        self.containers.push(ContainerContext {
+            annotations,
+            field_name,
+            container,
+        });
+        Ok(())
+    }
+
+    fn pop_container(&mut self) -> IonResult<ContainerContext> {
+        if self.containers.len() <= 1 {
+            return illegal_operation("cannot step out of the top level");
+        }
+        // `self.containers` is never empty; it always has at least the top level.
+        Ok(self.containers.pop().unwrap())
+    }
+
+    fn set_current_annotations(&mut self, annotations: Annotations) {
+        self.annotations = annotations;
+    }
+
+    fn write_scalar(&mut self, value: Value) -> IonResult<()> {
+        let annotations = self.take_current_annotations();
+        let element = value.with_annotations(annotations);
+        self.write_element(element)
+    }
+
+    fn write_element(&mut self, element: Element) -> IonResult<()> {
+        let field_name = self.field_name.take();
+        match &mut self.current_container_mut().container {
+            Container::TopLevel => self.output_mut().extend(std::iter::once(element)),
+            Container::SExpression(seq) => seq.push(element),
+            Container::List(seq) => seq.push(element),
+            Container::Struct(fields) => {
+                if let Some(field_name) = field_name {
+                    fields.push((field_name, element));
+                } else {
+                    return illegal_operation("Values inside a struct must have a field name.");
+                }
+            }
+        };
+        Ok(())
+    }
+
+    fn to_symbol<A>(&self, sym_ref: A) -> IonResult<Symbol>
+    where
+        A: AsRawSymbolTokenRef,
+    {
+        match sym_ref.as_raw_symbol_token_ref() {
+            RawSymbolTokenRef::SymbolId(_symbol_id) => {
+                // TODO what to do with symbol ids?
+                todo!()
+            }
+            RawSymbolTokenRef::Text(txt) => Ok(Symbol::owned(txt)),
+        }
+    }
+}
+
+impl<'e, E> IonWriter for ElementStreamWriter<'e, E>
+where
+    E: Extend<Element>,
+{
+    type Output = &'e mut E;
+
+    fn ion_version(&self) -> (u8, u8) {
+        (1, 0)
+    }
+
+    fn write_ion_version_marker(&mut self, _major: u8, _minor: u8) -> IonResult<()> {
+        // TODO what does this mean for an element writer?
+        todo!()
+    }
+
+    fn supports_text_symbol_tokens(&self) -> bool {
+        true
+    }
+
+    fn set_annotations<I, A>(&mut self, annotations: I)
+    where
+        A: AsRawSymbolTokenRef,
+        I: IntoIterator<Item = A>,
+    {
+        let annotations: IonResult<_> =
+            annotations.into_iter().map(|s| self.to_symbol(s)).collect();
+        // TODO what to do with the expect?
+        self.set_current_annotations(Annotations::new(annotations.expect("symbols")));
+    }
```

Comment:
This is one place that symbol ids could be passed to `to_symbol`, but this interface method is not fallible.

Reply:
Great catch. I went to see how the `BinaryWriter` handles this constraint and found an embarrassing leftover:

https://github.com/amazon-ion/ion-rust/blob/fd09e870e5206be16a1c535da0abdd450c6ea6b5/src/binary/binary_writer.rs#L124-L130

🤦

I've opened #564 to track addressing this down the line. For now, have it panic and link to the GH issue.